### PR TITLE
[+] match_string proved

### PR DESCRIPTION
**Pull request by SetSplin**

### CODE REVIEW

File: src/match_string.c [hunk 1]
```diff
@@ -1,66 +1,28 @@
...
-		"12345",
-		NULL
-	};
-
-	match_string(list, -1, str);
-	match_string(list, 3, str);
-
-	return 0;
-}
-#endif
```
Comment (evdenis) [inline src/match_string.c]: Please, keep the main function. Lines 44-66 should be kept.

File: src/match_string.h [hunk 2]
```diff
@@ -1,10 +1,8 @@
 #ifndef __MATCH_STRING_H__
 #define __MATCH_STRING_H__
 
-#include "kernel_definitions.h"
```
Comment (evdenis) [inline src/match_string.h]: What is the point of removing this include? Please, try to keep it.

File: src/match_string.h [hunk 3]
```diff
@@ -15,6 +13,38 @@
...
  */
 
+/*@
+    logic integer cmp(unsigned char a, unsigned char b) =
+       a == b ? 0 : a < b ? -1 : 1;
+
+    logic integer strncmp(char *cs, char *ct, integer n) =
+       n == -1 ? 0 : (cs[n] == ct[n] ? strncmp(cs+1, ct+1, n-1) : cmp((u8 AENO)cs[n], (u8 AENO)ct[n]));
+    logic integer strcmp(char *cs, char *ct) = strncmp(cs, ct, strlen(cs));
+*/
```
Comment (evdenis) [inline src/match_string.h]: These definitions should be included from strcmp.h file. Please, don't copy-paste them here.

File: src/match_string.h [hunk 4]
```diff
@@ -15,6 +13,38 @@
...
+
+/*@
+    requires \valid_read(array + (0..n-1));
+    requires valid_str(string);
+    requires n <= INT_MAX;
+    requires \forall integer i; 0 <= i < n && valid_str(array[i]);
```
Comment (evdenis) [inline src/match_string.h]: These requirements (28-29) are too strict.
According to the documentation n could be -1 ("@n:	number of strings in the array or -1 for NULL terminated arrays"). You need to reflect this in line 28. It's possible to have null in array, it should be reflected in 29.

File: src/match_string.h (hunk 3 above)
Comment (SetSplin) [reply src/match_string.h]: There is no @ symbol in comments in strcmp.h line 6. This causes an error "unbound function strcmp.". Should I add '@'?

**Review by evdenis — CHANGES_REQUESTED** [review]
You could add "proved" in README.md in the "Status" column for match_string.

File: src/match_string.h (hunk 3 above)
Comment (evdenis) [reply src/match_string.h]: Yes, you can edit these logic functions in strcmp.h

**Comment by evdenis** [comment]
Merged. Thanks!

**Pull request merged**